File: tree/101.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
    if root == nil {
        return true
    }
    return isSym(root.Left, root.Right)
}

func isSym(left *TreeNode, right *TreeNode) bool {
    if left == nil && right == nil {
        return true
    }
    if left != nil && right != nil {
        return left.Val == right.Val && isSym(left.Left, right.Right) && isSym(left.Right, right.Left)
    }
    return false
}

File: tree/101.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSymmetric(TreeNode* root) {
        return isSym(root->left, root->right);
    }
private:
    bool isSym(TreeNode *left, TreeNode *right) {
        if (!left && !right)
            return true;
        if (left && right)
            return left->val == right->val && isSym(left->left, right->right) && isSym(left->right, right->left);
        return false;
    }
};
